### PR TITLE
project: Add last job status as for project sync feedback feature.

### LINKED ISSUE #9872: No feedback when launching sync from details screen
##### ISSUE TYPE
 - Bug Report

##### SUMMARY
1. Create an inventory source, or a project
2. Go to the 'Details' view
3. See the 'Sync' button

Screenshot from 2021-04-09 17-35-43

Click it to start a sync.

There is no visible indicator that the sync was launched - no toast notification, no redirection to a view of the job, etc.

A user may click multiple times thinking nothing is happening, launching multiple syncs.

##### ENVIRONMENT
* AWX version: devel 12/10

### PR DESCRIPTION
**Pull request by seiwailai**
##### SUMMARY
Fixes #9872 

Add `last job status` as for the project sync feedback feature. When users clicked the `sync` button on the project page, `last job status` will eventually update the status to `Pending`, `Waiting`, `Running`, and final result (`Successful`, `Cancelled`, `Failed`, `Error`). The implementation requires the WebSocket connection to ensure synchronous status update.

In particular, the `last job status` is similar to the functionality of `status` in the project list (`status` in `ProjectListItem.jsx`). More specifically, the `last job status` accompanied by a job link that allowing the user to navigate to the job output page. Besides, there is also the tooltip that allowing the user to view the information related to the most recent sync, covering information like `JOB ID`, `STATUS`, `FINISHED`.

The rationale of having the `last job status` instead of redirection or toast notification is that:
1. User has choices on whether to redirect to the job output. If the user wishes to navigate to job output, he/she can click the link. Besides, the user might have other projects to be synced right after the current project and he/she may want to proceed back to the project list page instead of the job output page. If we implement force redirection, it would take a longer time to navigate to the project list page.
2. The status update on `last job status` is fundamentally similar to toast notification where the user can immediately be notified if he/she already clicked the `sync` button to launch the job.

Nevertheless, this PR requires further discussion. Any comments are welcomed!

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

 - UI
 
1. `awx/ui_next/src/screens/Project/Project.jsx`
2. `awx/ui_next/src/screens/Project/useWsProject.js` - Added websocket implementation
3. `awx/ui_next/src/screens/Project/ProjectDetail/ProjectDetail.jsx`

##### AWX VERSION

```
awx: 19.0.0
```

##### ADDITIONAL INFORMATION

In case if users spam the sync button, we will need to ensure the fluent UI on the most recent sync tooltip and last job status. Thus, we would not want to update our last job status to `Pending` if there is a current running job.

For instance, we clicked sync for a particular project twice. 
1. For the first sync, our last job status should immediately change to `Pending`, then `Waiting`, then `Running`, then result (which are `Successful`, `Failed`, `Error`, `Cancelled`).
2. For the second sync, if we have a running job, we should not update our UI to `Pending`, otherwise our most recent sync tooltip UI will lose our current running job and we cannot navigate to the job link through the link provided by `last job status` tooltip.

**Issue of `sync` button click spam**
Ideally, we should prevent any spamming on the `sync` button using backend logic to reduce overload on the server as we already have a similar running project. Together with backend logic, we can disable the sync button right after we start to sync a project.

However, if we only disable `sync` through the frontend, this seems insecure as people with bad intentions might able to change the button disable attribute.

**After**
project_sync_feedback

### CODE REVIEW

**Comment by nixocio** [comment]
Thanks for your PR. Is it ready to be reviewed?

**Comment by seiwailai** [comment]
> Thanks for your PR. Is it ready to be reviewed?

Yup, this is ready for review

File: awx/ui_next/src/screens/Project/useWsProject.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      // the fluent UI on most recent sync tooltip and last job status.
+      // Thus, we will not update our last job status to `Pending` if
+      // there is current running job.
+      //
+      // For instance, we clicked sync for particular project for twice.
+      // For first sync, our last job status should immediately change
+      // to `Pending`, then `Waiting`, then `Running`, then result
+      // (which are `successful`, `failed`, `error`, `cancelled`.
+      // For second sync, if the status response is `pending` and we have
+      // running and waiting jobs, we should not update our UI to `Pending`,
```
Comment (AlexSCorey) [inline awx/ui_next/src/screens/Project/useWsProject.js]:
```suggestion
      // running or waiting jobs, we should not update our UI to `Pending`,
```

Comment (AlexSCorey) [reply awx/ui_next/src/screens/Project/useWsProject.js]: We should also disable the sync button with a tooltip to explain why it is cancelled

**Review by AlexSCorey — CHANGES_REQUESTED** [review]
I really like this!  Just a couple of minor changes.  We should probably add some tests as well.

File: awx/ui_next/src/screens/Project/useWsProject.js (hunk 1 above)
Comment (seiwailai) [reply awx/ui_next/src/screens/Project/useWsProject.js]: > I really like this! Just a couple of minor changes. We should probably add some tests as well.

@AlexSCorey Thanks for your review, will make changes and add tests as soon as possible!

**Comment by jakemcdermott** [comment]
recheck

**Comment by mabashian** [comment]
@AlexSCorey is this just waiting for a re-review?

**Comment by seiwailai** [comment]
@AlexSCorey @nixocio @jakemcdermott This is ready for review!

**Review by AlexSCorey — APPROVED** [review]
Good well done!  Thanks for doing this.

**Review by tiagodread — CHANGES_REQUESTED** [review]
Hey @seiwailai thanks for adding all this context, this is awesome!
We have the same issue with the sync button span on the Project list, do you mind disabling the sync button on Project list as well?

**Comment by seiwailai** [comment]
> Hey @seiwailai thanks for adding all this context, this is awesome!
> We have the same issue with the sync button span on the Project list, do you mind disabling the sync button on Project list as well?

@tiagodread Sure thing!

**Comment by seiwailai** [comment]
> Hey @seiwailai thanks for adding all this context, this is awesome!
> We have the same issue with the sync button span on the Project list, do you mind disabling the sync button on Project list as well?

@tiagodread This is after changes for the project list!

project_list_sync_button

**Comment by tiagodread** [comment]
@seiwailai trying to create a project when I hit the save button the page is crashing:

chrome-capture (9)
image

**Comment by seiwailai** [comment]
@tiagodread I've added the conditional statement on rendering job status UI to solve the crash issue after project creation. Thank you for your comment!

**Newly-created project on sync**
add_project_on_sync

**Project on sync**
project_on_sync

**Project list on sync**
project_list_on_sync

**Review by tiagodread — APPROVED** [review]
Verified that:
- The user can create projects
- The sync button is disabled while the sync is running and the tooltip is displayed
- The user can click on the `Last Job Status` and navigate to the `Source Control Update` for the current project.
- All e2e tests are 🟢

**Pull request merged**